base/broadcast.jl

```diff
@@ -691,8 +691,52 @@ eltypes(t::Tuple{Any}) = Tuple{_broadcast_getindex_eltype(t[1])}
 eltypes(t::Tuple{Any,Any}) = Tuple{_broadcast_getindex_eltype(t[1]), _broadcast_getindex_eltype(t[2])}
 eltypes(t::Tuple) = Tuple{_broadcast_getindex_eltype(t[1]), eltypes(tail(t)).types...}
 
+function promote_typejoin_union(::Type{T}) where T
+    if T === Union{}
+        return Union{}
+    elseif T isa Union
+        return promote_typejoin(promote_typejoin_union(T.a), promote_typejoin_union(T.b))
+    elseif T <: Tuple
+        return typejoin_union_tuple(T)
+    else
+        return T
+    end
+end
+
+@pure function typejoin_union_tuple(T::Type)
+    u = Base.unwrap_unionall(T)
+    u isa Union && return typejoin(
+            typejoin_union_tuple(Base.rewrap_unionall(u.a, T)),
+            typejoin_union_tuple(Base.rewrap_unionall(u.b, T)))
+    p = (u::DataType).parameters
+    lr = length(p)::Int
+    if lr == 0
+        return Tuple{}
+    end
+    c = Vector{Any}(undef, lr)
+    for i = 1:lr
+        pi = p[i]
+        U = Core.Compiler.unwrapva(pi)
+        if U === Union{}
+            ci = Union{}
+        elseif U isa Union
+            ci = typejoin(Base.rewrap_unionall(U.a, T), Base.rewrap_unionall(U.b, T))
```

Comment:
@JeffBezanson is it okay that we're re-wrapping both these and the result in the same T?

Reply:
Hmm, right, we've hit something like this before (#36407). For example `Tuple{T, T, Ref{T}} where T`. If you rewrap `Ref{T}` then the outer `T` loses its invariant occurrence, making the other two Ts diagonal. So this can potentially give too narrow a type.

Reply:
I think `typejoin` tolerates types with free variables (at least, it calls itself recursively that way), so the fix is probably just to not call `rewrap` until the top level.

Reply:
In concrete terms, what should I do? :-D

Reply:
Replace this line with
```
ci = typejoin(U.a, U.b)
```